django/contrib/admin/filters.py

```diff
@@ -412,5 +412,13 @@ def choices(self, cl):
 
 class RelatedOnlyFieldListFilter(RelatedFieldListFilter):
     def field_choices(self, field, request, model_admin):
-        limit_choices_to = {'pk__in': set(model_admin.get_queryset(request).values_list(field.name, flat=True))}
-        return field.get_choices(include_blank=False, limit_choices_to=limit_choices_to)
+        limit_choices_to = {
+            'pk__in': (
+                model_admin
+                .get_queryset(request)
+                .distinct()
+                .values_list('{0}__pk'.format(self.field_path), flat=True)
```

Comment:
It also fix a wrong queryset result for ForeignKey with custom "to_field" attribute.